rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+        });
+        builder.append(StatisticsRow {
+            null_count: ScalarValue::from(0_i64),
+            min_value: ScalarValue::Int32(Some(std::i32::MIN)),
+            max_value: ScalarValue::Int32(Some(std::i32::MAX)),
+        });
+
+        // If we try to finish at this point, it will error since we don't have
+        // stats for b yet.
+        assert!(collector.finish().is_err());
+
+        // We cannot reuse old collector as it's builders were finished.
+        let mut collector = StatisticsCollector::new(&schema.fields);
+
+        let id = schema.field("a").unwrap().id;
+        let builder = collector.get_builder(id).unwrap();
+        builder.append(StatisticsRow {
+            null_count: ScalarValue::from(2_i64),
+            min_value: ScalarValue::from(1_i32),
+            max_value: ScalarValue::from(3_i32),
+        });
+        builder.append(StatisticsRow {
+            null_count: ScalarValue::from(0_i64),
+            min_value: ScalarValue::Int32(Some(std::i32::MIN)),
+            max_value: ScalarValue::Int32(Some(std::i32::MAX)),
+        });
+
+        // Collect stats for b
+        let id = schema.field("b").unwrap().id;
+        let builder = collector.get_builder(id).unwrap();
+        builder.append(StatisticsRow {
+            null_count: ScalarValue::from(6_i64),
+            min_value: ScalarValue::from("aaa"),
+            max_value: ScalarValue::from("bbb"),
+        });
+        builder.append(StatisticsRow {
+            null_count: ScalarValue::from(0_i64),
+            min_value: ScalarValue::Utf8(None),
+            max_value: ScalarValue::Utf8(None),
+        });
+
+        collector.append_num_values(42);
+        collector.append_num_values(64);
+
+        // Now we can finish
+        let batch = collector.finish().unwrap();
+
+        // TODO: Order of schema is not guaranteed and stable so this ocassionaly fails.
+        // Should we have a helper to compare batches?
+        let expected_schema = ArrowSchema::new(vec![
+            ArrowField::new("num_values", DataType::Int64, false),
+            ArrowField::new(
+                "1",
+                DataType::Struct(ArrowFields::from(vec![
+                    ArrowField::new("null_count", DataType::Int64, false),
+                    ArrowField::new("min_value", DataType::Utf8, true),
+                    ArrowField::new("max_value", DataType::Utf8, true),
+                ])),
+                false,
+            ),
+            ArrowField::new(
+                "0",
+                DataType::Struct(ArrowFields::from(vec![
+                    ArrowField::new("null_count", DataType::Int64, false),
+                    ArrowField::new("min_value", DataType::Int32, false),
+                    ArrowField::new("max_value", DataType::Int32, false),
+                ])),
+                false,
+            ),
+        ]);
+
+        assert_eq!(batch.schema().as_ref(), &expected_schema);
```

Comment:
It seems field order in schema is not stable and this check will occasionally fail. Is this expected and should we provide a helper for such test cases?

Reply:
That's because `builders` in `StatisticsCollector` is a `HashMap`, which doesn't preserve order. Perhaps you can change it to a `BTreeMap`, which is ordered? Or when `finish()` is called it can sort the builders by `field_id` before creating the final batch?